support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -233,3 +250,81 @@ object CreateZuoraSubscription {
     }
 
 }
+
+object DigitalSubscriptionGiftRedemption {
+
+  def ifDigitalSubscriptionGiftRedemption(product: ProductType, paymentMethod: Either[PaymentMethod, RedemptionData]) = {
+    product match {
+      case d: DigitalPack if paymentMethod.isRight && d.readerType == Gift => paymentMethod.right.toOption
+      case _ => None
+    }
+  }
+
+  def redeemGift(
+    redemptionData: RedemptionData,
+    gifteeUserId: String,
+    requestInfo: RequestInfo,
+    state: CreateZuoraSubscriptionState,
+    zuoraService: ZuoraService,
+    catalogService: CatalogService
+  ): Future[HandlerResult[SendThankYouEmailState]] = {
+    for {
+      giftSubscriptionFields <- zuoraService.getSubscriptionFromRedemptionCode(redemptionData.redemptionCode)
+      validatedSubscription <- Future.fromTry(validateRedemptionData(giftSubscriptionFields))
+      fullGiftSubscription <- zuoraService.getSubscriptionById(validatedSubscription.id)
+      newTermLength <- Future.fromTry(calculateNewTermLength(fullGiftSubscription, catalogService))
+      updateDataResponse <- zuoraService.updateSubscriptionRedemptionData(validatedSubscription.id, gifteeUserId, newTermLength)
+      handlerResult <- Future.fromTry(buildHandlerResult(updateDataResponse, state, redemptionData, requestInfo))
+    } yield handlerResult
+  }
+
+  def validateRedemptionData(existingSub: SubscriptionRedemptionQueryResponse) = {
+    existingSub.records.headOption.map(
+      existingSubFields =>
+        if (existingSubFields.gifteeIdentityId.isEmpty)
+          Success(existingSubFields)
+        else
+          Failure(new RuntimeException(GetCodeStatus.CodeAlreadyUsed.clientCode))
+    ).getOrElse(Failure(new RuntimeException(GetCodeStatus.NoSuchCode.clientCode)))
```

Comment:
minor but not sure we need to use the client code here, as it's not needed to be machine readable, this will just be in the support workers failure.  Just toStringing the status object should be okay and clear enough to humans.

Reply:
I'm checking the message in an IT test so it's nice to have a val to compare it to

Reply:
fair enough 👍